magetypes: piecewise tier-split blocks impossible — _scalar auto-emitted from every block

## Summary

The natural pattern for splitting per-tier-natural SIMD widths across two `#[magetypes]` blocks doesn't compile because the resolver unconditionally auto-appends a `_scalar` variant to any block whose tier list lacks `scalar` or `default`, causing both blocks to emit `_scalar` and collide.

## Repro

```rust
#[magetypes(define(f32x16), v4, v4x)]
fn foo(token: Token, x: &mut [f32]) {
    /* AVX-512 native f32x16 body */
}

#[magetypes(define(f32x8), v3, neon, wasm128)]
fn foo(token: Token, x: &mut [f32]) {
    /* AVX2 / NEON / WASM SIMD128 native f32x8 body */
}
```

```
error[E0428]: the name `foo_scalar` is defined multiple times
```

## Root cause

`archmage-macros-0.9.22/src/tiers.rs:397-403`:

```rust
if !has_scalar && !has_default {
    tiers.push(ResolvedTier {
        tier: find_tier("scalar").unwrap(),
        ...
    });
}
```

This unconditional auto-append makes the documented "piecewise tier blocks" pattern impossible without falling back to standalone `#[arcane]` overrides (Pattern C), which forfeits the macro's `define(...)` ergonomics for the hand-written tier.

## Workarounds tried (none work)

- `#[magetypes(define(f32x16), v4, v4x, default)]`: `default` is tokenless, body's `Token` ident becomes undefined.
- `#[magetypes(define(f32x16), +v4, +v4x, -v3, -neon, -wasm128, -scalar)]`: additive mode resolves correctly but the auto-append at line 397 *re-adds* `scalar` after the `-scalar` removal.
- `#[magetypes(define(f32x16), v4, v4x)]` with `#[cfg(target_arch = "x86_64")]` guard: cfg propagates to all variants including the auto-emitted `_scalar`, but the second block's `_scalar` has no cfg and they still collide on x86_64.

## Suggested fixes (any one)

1. **Skip the auto-append when the user's tier list is non-empty.** The fallback only protects users who write `#[magetypes()]` with no tiers at all. If they listed *any* tiers, presumably they know which ones they want.
2. **Honor `-scalar` in non-additive mode** as an explicit "don't auto-append" signal. Currently the parser refuses to mix `+`/`-` with plain names (tiers.rs:287-292).
3. **Add an explicit `no_scalar` / `no_fallback` flag** alongside `define(...)` and `rite`.

(1) is the smallest change and matches the principle of least surprise — "I listed v4 and v4x, that's what I want."

## Use case context

Splitting tiers by natural SIMD width matters for hot loops that benefit from the polyfill being avoided on a particular tier. In zensim's `downscale_2x_into`, AVX-512 wants f32x16 native and AVX2 wants f32x8 native; using one block with f32x16 across all tiers polyfills to 2× f32x8 on AVX2, costing ~5-15% on Zen 4 v3 dispatch. Two blocks would express this naturally.

Currently we work around with one `#[magetypes(...)]` block + standalone `#[arcane]` for the lone outlier (Pattern C), which works but loses the `define(f32x8)` ergonomic for the v3 body.

cc @lilith

--- a/archmage-macros/src/lib.rs
+++ b/archmage-macros/src/lib.rs
@@ -1082,6 +1082,27 @@ mod tests {
     }
 
     #[test]
+    fn resolve_subtractive_removes_scalar() {
+        // -scalar must actually remove scalar — auto-append must not undo it
+        let tiers = resolve_tier_names(&["-scalar"], true);
+        assert!(
+            !tiers.iter().any(|t| t == "scalar"),
+            "expected scalar to be removed, got {tiers:?}"
+        );
+    }
+
+    #[test]
+    fn resolve_subtractive_removes_scalar_with_other_modifiers() {
+        // [-scalar, +arm_v2] — scalar removed, arm_v2 added, no scalar re-injected
+        let tiers = resolve_tier_names(&["-scalar", "+arm_v2"], true);
+        assert!(
+            !tiers.iter().any(|t| t == "scalar"),
+            "expected scalar to be removed, got {tiers:?}"
+        );
+        assert!(tiers.contains(&"arm_v2".to_string()));
+    }
+
+    #[test]
     fn resolve_mixed_add_remove() {
         let tiers = resolve_tier_names(&["-neon", "-wasm128", "+v1"], true);
         assert!(tiers.contains(&"v1".to_string()));

--- a/archmage-macros/src/magetypes.rs
+++ b/archmage-macros/src/magetypes.rs
@@ -22,7 +22,9 @@ use crate::tiers::*;
 /// `defines` is a list of magetypes type names (e.g. `["f32x8", "u16x16"]`)
 /// to inject as local type aliases at the top of each variant's body:
 ///
-///     type f32x8 = ::magetypes::simd::generic::f32x8<Token>;
+/// ```text
+/// type f32x8 = ::magetypes::simd::generic::f32x8<Token>;
+/// ```
 ///
 /// The alias's `Token` is substituted to the concrete token type for each
 /// tier (same as the rest of the body). This eliminates the boilerplate

--- a/archmage-macros/src/tiers.rs
+++ b/archmage-macros/src/tiers.rs
@@ -303,6 +303,10 @@ pub(crate) fn resolve_tiers(
     // re-apply the descriptor's cfg_feature on them. Writing +v4 means "I want
     // v4 exactly as written, without the automatic avx512 gate."
     let mut user_overrides: Vec<String> = Vec::new();
+    // Tracks whether the user explicitly removed the scalar/default fallback
+    // via `-scalar` or `-default`. Used to suppress the auto-append below so
+    // that `[-scalar]` actually drops the scalar variant.
+    let mut fallback_explicitly_removed = false;
     let effective_names: Vec<String> = if any_modifier {
         let mut names: Vec<String> = DEFAULT_TIER_NAMES.iter().map(|s| s.to_string()).collect();
         for raw in tier_names {
@@ -330,6 +334,9 @@ pub(crate) fn resolve_tiers(
             if is_removal {
                 if let Some(pos) = pos {
                     names.remove(pos);
+                }
+                if is_fallback {
+                    fallback_explicitly_removed = true;
                 }
                 // Removing a tier that's not in defaults is a silent no-op
             } else if let Some(pos) = pos {
@@ -394,7 +401,7 @@ pub(crate) fn resolve_tiers(
         ));
     }
 
-    if !has_scalar && !has_default {
+    if !has_scalar && !has_default && !fallback_explicitly_removed {
         tiers.push(ResolvedTier {
             tier: find_tier("scalar").unwrap(),
             feature_gate: None,

